src/hotspot/share/opto/predicates.cpp

```diff
@@ -381,6 +382,83 @@ bool TemplateAssertionPredicateExpressionNode::is_template_assertion_predicate(N
+//                    If                                             new If
+//                  /    \                                           /    \
+//            success     fail path                         new success   new Halt
+//              proj    (Halt or UCT)                           proj
+//
+IfTrueNode* InitializedAssertionPredicate::create(Node* control) {
+  IdealLoopTree* loop = _phase->get_loop(control);
+  OpaqueInitializedAssertionPredicateNode* new_opaque_bool = create_new_bool(control);
+  IfNode* if_node = create_if_node(control, new_opaque_bool, loop);
+  create_halt_path(if_node, loop);
+  return create_success_path(if_node, loop);
+}
+
+// Create a OpaqueInitializedAssertionPredicate with a new Assertion Predicate Expression.
+OpaqueInitializedAssertionPredicateNode* InitializedAssertionPredicate::create_new_bool(Node* control) {
+  Opaque4Node* template_opaque_bool = _template_assertion_predicate->in(1)->as_Opaque4();
+  TemplateAssertionPredicateExpression template_assertion_predicate_expression(template_opaque_bool);
+  Opaque4Node* new_opaque =
+      template_assertion_predicate_expression.clone_and_replace_init_and_stride(_new_init, _new_stride, control, _phase);
+  OpaqueInitializedAssertionPredicateNode* opaque_bool
+      = new OpaqueInitializedAssertionPredicateNode(new_opaque->in(1)->as_Bool(), _phase->C);
+  _phase->register_new_node(opaque_bool, control);
+  return opaque_bool;
+}
+
+IfNode* InitializedAssertionPredicate::create_if_node(Node* control,
+                                                      OpaqueInitializedAssertionPredicateNode* new_opaque_bool,
+                                                      IdealLoopTree* loop) {
+  const int if_opcode = _template_assertion_predicate->Opcode();
+  IfNode* if_node = if_opcode == Op_If ?
+      new IfNode(control, new_opaque_bool, PROB_MAX, COUNT_UNKNOWN NOT_PRODUCT(COMMA _assertion_predicate_type)) :
+      new RangeCheckNode(control, new_opaque_bool, PROB_MAX, COUNT_UNKNOWN NOT_PRODUCT(COMMA _assertion_predicate_type));
+  _phase->register_control(if_node, loop, control);
+  return if_node;
+}
+
+IfTrueNode* InitializedAssertionPredicate::create_success_path(IfNode* if_node, IdealLoopTree* loop) {
+  IfTrueNode* success_proj = new IfTrueNode(if_node);
+  _phase->register_control(success_proj, loop, if_node);
+  return success_proj;
+}
+
+void InitializedAssertionPredicate::create_halt_path(IfNode* if_node, IdealLoopTree* loop) {
+  IfFalseNode* fail_proj = new IfFalseNode(if_node);
+  _phase->register_control(fail_proj, loop, if_node);
+  create_halt_node(fail_proj, loop);
+}
+
+void InitializedAssertionPredicate::create_halt_node(IfFalseNode* fail_proj, IdealLoopTree* loop) {
+  StartNode* start_node = _phase->C->start();
+  Node* frame = new ParmNode(start_node, TypeFunc::FramePtr);
+  _phase->register_new_node(frame, start_node);
+  Node* halt = new HaltNode(fail_proj, frame, "Assertion Predicate cannot fail");
```

Comment:
```suggestion
  Node* halt = new HaltNode(fail_proj, frame, "Initialized Assertion Predicate cannot fail");
```
I would add the extra detail, just to have more context when it fails ;)